java/src/main/native/include/jni_utils.hpp

```diff
@@ -863,21 +895,26 @@ inline void jni_cuda_check(JNIEnv *const env, cudaError_t cuda_status) {
     JNI_CHECK_CUDA_ERROR(env, cudf::jni::CUDA_ERROR_CLASS, e, ret_val);                            \
   }                                                                                                \
   catch (const cudf::data_type_error &e) {                                                         \
-    JNI_CHECK_THROW_NEW(env, cudf::jni::CUDF_DTYPE_ERROR_CLASS, e.what(), ret_val);                \
+    JNI_CHECK_THROW_NEW_CUDF(env, cudf::jni::CUDF_DTYPE_ERROR_CLASS, e.what(), e.stacktrace(),     \
+                             ret_val);                                                             \
   }                                                                                                \
   catch (const std::exception &e) {                                                                \
+    char const *ex_class = class_name;                                                             \
     /* Double check whether the thrown exception is unrecoverable CUDA error or not. */            \
     /* Like cudf::detail::throw_cuda_error, it is nearly certain that a fatal error  */            \
     /* occurred if the second call doesn't return with cudaSuccess. */                             \
     cudaGetLastError();                                                                            \
     auto const last = cudaFree(0);                                                                 \
     if (cudaSuccess != last && last == cudaDeviceSynchronize()) {                                  \
-      auto msg = e.what() == nullptr ? std::string{""} : e.what();                                 \
-      auto cuda_error = cudf::fatal_cuda_error{msg, last};                                         \
-      JNI_CHECK_CUDA_ERROR(env, cudf::jni::CUDA_FATAL_ERROR_CLASS, cuda_error, ret_val);           \
+      /* Throw CudaFatalException since the thrown exception is unrecoverable CUDA error */        \
+      ex_class = cudf::jni::CUDA_FATAL_ERROR_CLASS;                                                \
```

Comment:
Oh no, I was wrong with this. Should call `JNI_CHECK_CUDA_ERROR` to record the cuda error code. Working on this now...

Reply:
Done.